Convert tests using enzyme to use react-test-renderer instead (components/inserter/test/menu.js)

nerrad opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

This pull converts all usage of enzyme for tests in `components/inserter/test/menu.js` to use `react-test-renderer` instead.  This is because enzyme does not support React 16.3+ (and movement to do so is really slow).  This will fix issues with breakage due to the enzyme incompatibility as components receive React 16.3+ features (such as `forwardRef` usage in #7557)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

nerrad committed:
convert enzyme usage to react-test-renderer usage.

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -107,12 +139,21 @@ describe( 'InserterMenu', () => {
-		const activeCategory = wrapper.find( '.components-panel__body.is-opened > .components-panel__body-title' );
-		expect( activeCategory.text() ).toBe( 'Most Used' );
+		const activeCategory = wrapper.root.find( ( node ) => {
+			return node.props.className &&
```
- When will `className` be falsey?
- This is more a perfect use case for [`classList.contains`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) (assuming `jsdom` supports it)

aduth replied:
Also this is prone to two issues:

- If order changes between application of `is-opened` and `components-panel__body`
- If suffixes/prefixes are applied to the class names we're testing (e.g. `not-the-components-panel__body is-opened-jk`)

nerrad replied:
> When will className be falsey?

`find` traverses ALL components to verify there is only one match (throws an error as soon as there is more than one).  So it's possible it hits a component where `props.className` is undefined.

Good points on your subsequent comment though that need addressed.

nerrad replied:
> This is more a perfect use case for classList.contains (assuming jsdom supports it)

`classList` is not exposed on the `node` in this iterator.  I'm a bit unclear on what you had in mind here.  Were you thinking of a different approach instead of using `TestRenderer.root.find`?

nerrad replied:
> False negatives if order changes between application of is-opened and components-panel__body
> False positives if suffixes/prefixes are applied to the class names we're testing (e.g. not-the-components-panel__body is-opened-jk)

On re-examination, didn't those issues exist in the original test?  I'm all for improving tests but the scope of this pull was a straight conversion.  I'm open to also improving the test if thats the desire.

aduth replied:
Sorry, I mistook this as an HTMLElement, when in fact it's the React node. 

>On re-examination, didn't those issues exist in the original test? I'm all for improving tests but the scope of this pull was a straight conversion.

I don't think so. With the following diff applied, I'd expect the current tests to fail while the previous would have passed:

```diff
diff --git a/components/panel/body.js b/components/panel/body.js
index 1a22de279..048c3193d 100644
--- a/components/panel/body.js
+++ b/components/panel/body.js
@@ -40,7 +40,7 @@ class PanelBody extends Component {
 		const { title, children, opened, className, icon } = this.props;
 		const isOpened = opened === undefined ? this.state.opened : opened;
 		const arrow = `arrow-${ isOpened ? 'up' : 'down' }`;
-		const classes = classnames( 'components-panel__body', className, { 'is-opened': isOpened } );
+		const classes = classnames( { 'is-opened': isOpened }, 'components-panel__body', className );
 
 		return (
 			<div className={ classes }>
```

aduth replied:
Well, it looks like the tests were since updated, but in the original implementation it would have been an issue.

aduth replied:
I expect the following would still result in a false positive:

```diff
diff --git a/components/panel/body.js b/components/panel/body.js
index 1a22de279..65a86abeb 100644
--- a/components/panel/body.js
+++ b/components/panel/body.js
@@ -40,7 +40,7 @@ class PanelBody extends Component {
 		const { title, children, opened, className, icon } = this.props;
 		const isOpened = opened === undefined ? this.state.opened : opened;
 		const arrow = `arrow-${ isOpened ? 'up' : 'down' }`;
-		const classes = classnames( 'components-panel__body', className, { 'is-opened': isOpened } );
+		const classes = classnames( 'components-panel__body', className, { 'is-opened-not': isOpened } );
 
 		return (
 			<div className={ classes }>
```

nerrad replied:
yup it would. revisiting.

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -107,12 +139,21 @@ describe( 'InserterMenu', () => {
+			return node.props.className &&
+				node.props.className.includes( 'components-panel__body-title' );
+		} ).find( ( node ) => {
+			return node.type === 'button';
```
Oof, this is a fair bit of effort to find a descendent node. I wonder if we could do for some of our small abstractions for common use cases.

nerrad replied:
Another option here would be to convert the component to a DOM node and then use `querySelector` on it.  That will at least reduce this traversal to a single selector string similar to what was originally used.

nerrad committed:
account for possibility of class name string being in different order

nerrad committed:
update tests to use TestUtils

This allows for more accurately testing rendered expectations.

nerrad commented:
@aduth I rewrote these tests to use `react-dom/test-utils` instead.  This conversion more closely preserves the original tests and addresses the feedback you gave here.  However its a great example of a test that probably should just be an e2e test.  In the interest of keeping things moving forward however, I hope its okay to accept this conversion on an _existing_ test.

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -90,256 +91,240 @@ const items = [
+
+const initializeMenuDefaultStateAndReturnElement = ( propOverrides ) => {
+	setWrapperForProps( propOverrides );
+	/* eslint-disable react/no-find-dom-node */
```
We should consider overriding test files in ESLint to allow for `findDOMNode`

https://eslint.org/docs/user-guide/configuring#disabling-rules-only-for-a-group-of-files

nerrad replied:
> We should consider overriding test files in ESLint to allow for findDOMNode

That would be good, should be done in a separate pull though?

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -90,256 +91,240 @@ const items = [
-	// NOTE: Due to https://github.com/airbnb/enzyme/issues/1174, some of the selectors passed through to
-	// wrapper.find have had to be strengthened (and the filterWhere strengthened also), otherwise two
-	// results would be returned even though only one was in the DOM.
+const defaultProps = {
```
This would fall under `SCREAMING_SNAKE_CASE` constant convention:

https://github.com/WordPress/gutenberg/blob/master/docs/reference/coding-guidelines.md#constants

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -90,256 +91,240 @@ const items = [
+	setTimeout: noop,
+};
+
+let wrapper;
```
Doesn't need to be module scoped? Should this be the return value of `setWrapperForProps`, then `initializeMenuDefaultStateAndReturnElement` can internally assign it to a `wrapper` variable if need be?

aduth replied:
Might also make `setWrapperForProps` redundant altogether.

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -90,256 +91,240 @@ const items = [
+
+let wrapper;
+
+const setWrapperForProps = ( propOverrides ) => {
```
Honest curiosity: Why arrow functions vs. the `function` keyword?

nerrad replied:
habit mostly :)

aduth reviewed (approved):
Nice 👍

nerrad committed:
fixes from review

- use SCREAMING_SNAKECASE for constant (standards)
- refactor unnecessary module scoped variable

aduth commented on editor/components/inserter/test/menu.js:
```diff
@@ -90,256 +91,238 @@ const items = [
-	// results would be returned even though only one was in the DOM.
+const DEFAULT_PROPS = {
+	position: 'top center',
+	instanceID: 1,
```
`withInstanceId` passes a prop called `instanceId`, not `instanceID`. Should we be concerned that this is not failing? Or do we not need it at all?

nerrad replied:
ooo good catch!  I wouldn't expect things to fail in the _current_ tests because `instanceId` is only implemented on the search label `for` attribute and the search input `id` attribute.  We're not testing those rendered attributes specifically.

Still, this should probably be fixed by either removing instanceID from the props or just changing it to `instanceId`.  I would prefer removing it.  I also only think this should be fixed to prevent confusion for other devs looking at this test, but really its not impacting existing tests (similar to how the lack of a `rootUID` passed in as props on tests isn't affecting things. 

Fix in a new pull?

aduth replied:
Yep, new pull sounds great. If we don't need it, removing it sounds best.

nerrad replied:
Somewhat related.  I wonder if these components should have defaultProps for implemented props that could be undefined if the component isn't wrapped (eg. `instanceId` or `rootUID`?  Or is the component just exported for convenience with testing?

aduth replied:
>Or is the unwrapped component just exported for convenience with testing?

Primarily, yes.

nerrad replied:
https://github.com/WordPress/gutenberg/pull/7984
